internal/scheduler/internaltypes/resource_list.go

```diff
@@ -1,16 +1,196 @@
 package internaltypes
 
-import "fmt"
+import (
+	"fmt"
+
+	"golang.org/x/exp/slices"
+	k8sResource "k8s.io/apimachinery/pkg/api/resource"
+)
 
 type ResourceList struct {
-	resources []int64
-	factory   *ResourceListFactory
+	resources []int64              // immutable, do not change this, return a new struct instead!
+	factory   *ResourceListFactory // immutable, do not change this!
+}
+
+type Resource struct {
+	Name  string
+	Value int64
+	Scale k8sResource.Scale
 }
 
-func (rl *ResourceList) GetByName(name string) (int64, error) {
+func (rl ResourceList) Equal(other ResourceList) bool {
+	if rl.IsEmpty() && other.IsEmpty() {
+		return true
+	}
+	if rl.IsEmpty() || other.IsEmpty() {
+		return false
+	}
+	return slices.Equal(rl.resources, other.resources)
+}
+
+func (rl ResourceList) String() string {
+	if rl.IsEmpty() {
+		return "empty"
+	}
+	result := ""
+	for i, name := range rl.factory.indexToName {
+		if i > 0 {
+			result += " "
+		}
+		result += fmt.Sprintf("%s=%s", name, rl.asQuantity(i).String())
+	}
+	return result
+}
+
+func (rl ResourceList) GetByName(name string) (int64, error) {
+	if rl.IsEmpty() {
+		return 0, fmt.Errorf("resource type %s present as resource list is nil", name)
+	}
 	index, ok := rl.factory.nameToIndex[name]
 	if !ok {
 		return 0, fmt.Errorf("resource type %s not found", name)
 	}
 	return rl.resources[index], nil
 }
+
+func (rl ResourceList) GetByNameZeroIfMissing(name string) int64 {
+	if rl.IsEmpty() {
+		return 0
+	}
+	index, ok := rl.factory.nameToIndex[name]
+	if !ok {
+		return 0
+	}
+	return rl.resources[index]
+}
+
+func (rl ResourceList) GetResources() []Resource {
+	if rl.IsEmpty() {
+		return []Resource{}
+	}
+
+	result := make([]Resource, len(rl.resources))
+	for i, q := range rl.resources {
+		result[i] = Resource{
+			Name:  rl.factory.indexToName[i],
+			Value: q,
+			Scale: rl.factory.scales[i],
+		}
+	}
+	return result
+}
+
+func (rl ResourceList) AllZero() bool {
+	if rl.IsEmpty() {
+		return true
+	}
+	for _, r := range rl.resources {
+		if r != 0 {
+			return false
+		}
+	}
+	return true
+}
+
+func (rl ResourceList) HasNegativeValues() bool {
+	if rl.IsEmpty() {
+		return false
+	}
+	for _, r := range rl.resources {
+		if r < 0 {
+			return true
+		}
+	}
+	return false
+}
+
+func (rl ResourceList) IsEmpty() bool {
+	return rl.factory == nil
+}
+
+func (rl ResourceList) ExceedsAvailable(available ResourceList) (string, k8sResource.Quantity, k8sResource.Quantity, bool) {
```

Comment:
I'd be tempted to document this function as it's non trivial (e.g. it returns three values)

Reply:
Fixed